File: app/src/main/java/com/borisruzanov/russianwives/ui/fragments/ActionsFragment.java
package com.borisruzanov.russianwives.ui.fragments;


import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.arellomobile.mvp.MvpAppCompatFragment;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.borisruzanov.russianwives.Adapters.ActionsAdapter;
import com.borisruzanov.russianwives.OnItemClickListener;
import com.borisruzanov.russianwives.R;
import com.borisruzanov.russianwives.models.ActionItem;
import com.borisruzanov.russianwives.mvp.model.interactor.ActionsInteractor;
import com.borisruzanov.russianwives.mvp.model.repository.FirebaseRepository;
import com.borisruzanov.russianwives.mvp.presenter.ActionsPresenter;
import com.borisruzanov.russianwives.mvp.view.ActionsView;
import com.borisruzanov.russianwives.ui.FriendProfileActivity;

import java.util.List;

public class ActionsFragment extends MvpAppCompatFragment implements ActionsView {

  @InjectPresenter
  ActionsPresenter actionsPresenter;

  @ProvidePresenter
  public ActionsPresenter provideActionsPresenter(){
    return new ActionsPresenter(new ActionsInteractor(new FirebaseRepository()));
  }

  RecyclerView recyclerActivitiesList;
  ActionsAdapter actionsAdapter;

  private TextView emptyText;

  public ActionsFragment() {
    // Required empty public constructor
  }


  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {

    View view = inflater.inflate(R.layout.fragment_main_tab_activities, container, false);

    recyclerActivitiesList = view.findViewById(R.id.friends_fragment_recycler_activities);
    recyclerActivitiesList.setLayoutManager(new LinearLayoutManager(getActivity()));
    recyclerActivitiesList.setHasFixedSize(true);
    recyclerActivitiesList.addItemDecoration(new DividerItemDecoration(recyclerActivitiesList.getContext(),
            DividerItemDecoration.VERTICAL));

    actionsAdapter = new ActionsAdapter(onItemClickCallback);
    recyclerActivitiesList.setAdapter(actionsAdapter);

    emptyText = view.findViewById(R.id.activities_empty_text);

    //actionsPresenter.setActionsList();

    // Inflate the layout for this fragment
    return view;
  }

  @Override
  public void showUserActions(List<ActionItem> actionItems) {
    if(!actionItems.isEmpty()) {
      emptyText.setVisibility(View.GONE);
      recyclerActivitiesList.post(() -> actionsAdapter.setData(actionItems));
    }
    else {
      recyclerActivitiesList.setVisibility(View.GONE);
      emptyText.setVisibility(View.VISIBLE);
    }
  }

  private OnItemClickListener.OnItemClickCallback onItemClickCallback = (view, position) ->
          actionsPresenter.openFriendProfile(position);

  @Override
  public void openFriendProfile(String friendUid){
    Intent friendProfileIntent = new Intent(getActivity(), FriendProfileActivity.class);
    friendProfileIntent.putExtra("uid", friendUid);
    startActivity(friendProfileIntent);
  }


}

File: app/src/main/java/com/borisruzanov/russianwives/mvp/ui/actions/ActionsFragment.kt
package com.borisruzanov.russianwives.mvp.ui.actions


import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import com.arellomobile.mvp.MvpAppCompatFragment
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.borisruzanov.russianwives.OnItemClickListener
import com.borisruzanov.russianwives.R
import com.borisruzanov.russianwives.di.component
import com.borisruzanov.russianwives.models.ActionItem
import com.borisruzanov.russianwives.mvp.ui.actions.adapter.ActionsAdapter
import com.borisruzanov.russianwives.mvp.ui.friendprofile.FriendProfileActivity
import javax.inject.Inject

class ActionsFragment : MvpAppCompatFragment(), ActionsView {

    @Inject
    @InjectPresenter
    lateinit var actionsPresenter: ActionsPresenter

    private lateinit var recyclerActivitiesList: RecyclerView

    private lateinit var emptyLayout: RelativeLayout

    private val onItemClickCallback
            = OnItemClickListener.OnItemClickCallback{ _: View, position: Int -> actionsPresenter.openFriendProfile(position) }

    private val actionsAdapter = ActionsAdapter(onItemClickCallback)

    @ProvidePresenter
    fun provideActionsPresenter(): ActionsPresenter = actionsPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        requireActivity().component.inject(this)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val view = inflater.inflate(R.layout.fragment_main_tab_activities, container, false)

        // todo: use syntetic
        recyclerActivitiesList = view.findViewById(R.id.friends_fragment_recycler_activities)
        recyclerActivitiesList.layoutManager = LinearLayoutManager(activity)
        recyclerActivitiesList.setHasFixedSize(true)
        recyclerActivitiesList.addItemDecoration(DividerItemDecoration(recyclerActivitiesList.context,
                DividerItemDecoration.VERTICAL))

        recyclerActivitiesList.adapter = actionsAdapter

        emptyLayout = view.findViewById(R.id.actions_empty_rl)

        // Inflate the layout for this fragment
        return view
    }

    override fun showUserActions(actionItems: List<ActionItem>) {
        if (actionItems.isNotEmpty()) {
            emptyLayout.visibility = View.GONE
            recyclerActivitiesList.post { actionsAdapter.setData(actionItems) }
        } else {
            recyclerActivitiesList.visibility = View.GONE
            emptyLayout.visibility = View.VISIBLE
        }
    }

    override fun updateUserActions(newActionItems: List<ActionItem>) {
        if (newActionItems.isNotEmpty()) {
            emptyLayout.visibility = View.GONE
            if (actionsAdapter.itemCount == 0){
                recyclerActivitiesList.post { actionsAdapter.clearAndUpdateData(newActionItems) }
            }
        } else {
            recyclerActivitiesList.visibility = View.GONE
            emptyLayout.visibility = View.VISIBLE
        }
    }

    override fun openFriendProfile(friendUid: String) {
        val friendProfileIntent = Intent(activity, FriendProfileActivity::class.java)
        friendProfileIntent.putExtra("uid", friendUid)
        startActivity(friendProfileIntent)
    }

    override fun onResume() {
        super.onResume()

        actionsPresenter.updateActionsList()
    }

}